proof/invariant-abstract/Syscall_AI.thy

```diff
@@ -45,37 +45,89 @@ lemmas [wp] =
 lemmas [simp] =
   data_to_cptr_def
 
-crunch ethread_get, ethread_get_when
-  for inv[wp]: P
 
-lemma schedule_invs[wp]: "\<lbrace>invs\<rbrace> (Schedule_A.schedule :: (unit,det_ext) s_monad) \<lbrace>\<lambda>rv. invs\<rbrace>"
-  supply if_split[split del]
-  apply (simp add: Schedule_A.schedule_def)
-  apply (wp dmo_invs thread_get_inv gts_wp
-            do_machine_op_tcb when_def hoare_vcg_all_lift
-          | wpc
-          | clarsimp simp: guarded_switch_to_def get_tcb_def choose_thread_def ethread_get_def
-                           ethread_get_when_def
-          | wp (once) hoare_drop_imps
-          | simp add: schedule_choose_new_thread_def if_apply_def2)+
+crunch next_domain
+  for scheduler_action[wp]: "\<lambda>s. P (scheduler_action s)"
+  and valid_idle[wp]: valid_idle
+  (simp: crunch_simps)
+
+lemma next_domain_invs[wp]:
+  "next_domain \<lbrace> invs \<rbrace>"
+  unfolding next_domain_def
+  apply (wpsimp simp: Let_def)
+  apply (simp add: invs_def cur_tcb_def valid_state_def
+                   valid_mdb_def mdb_cte_at_def valid_ioc_def valid_irq_states_def
+                   valid_machine_state_def)
   done
 
-lemma schedule_choose_new_thread_ct_activatable[wp]:
-  "\<lbrace> invs \<rbrace> schedule_choose_new_thread \<lbrace>\<lambda>_. ct_in_state activatable \<rbrace>"
-  proof -
+crunch tcb_sched_action
+  for cur_thread[wp]: "\<lambda>s. P (cur_thread s)"
+  and scheduler_action[wp]: "\<lambda>s. P (scheduler_action s)"
+  (simp: crunch_simps)
+
+lemma cur_thread_update_invs:
+  "\<lbrace>\<lambda>s. invs s \<and> scheduler_action s \<noteq> resume_cur_thread \<and> tcb_at tptr s\<rbrace>
+   modify (cur_thread_update (\<lambda>_. tptr))
+   \<lbrace>\<lambda>_. invs\<rbrace>"
+  by (wpsimp simp: invs_def valid_state_def valid_pspace_def valid_machine_state_def
+                   state_refs_of_def cur_tcb_def)
+
+lemma switch_to_thread_invs:
+  "\<lbrace>\<lambda>s. invs s \<and> scheduler_action s \<noteq> resume_cur_thread\<rbrace>
+   switch_to_thread tptr
+   \<lbrace>\<lambda>_. invs\<rbrace>"
+  by (wpsimp simp: switch_to_thread_def thread_get_def is_tcb
+               wp: cur_thread_update_invs)
+
+lemma guarded_switch_to_invs:
+  "\<lbrace>\<lambda>s. invs s \<and> scheduler_action s \<noteq> resume_cur_thread\<rbrace>
+   guarded_switch_to thread
+   \<lbrace>\<lambda>_. invs\<rbrace>"
+  by (wpsimp simp: guarded_switch_to_def
+               wp: switch_to_thread_invs hoare_drop_imps)
+
+(* still true without scheduler_action s \<noteq> resume_cur_thread, but the proof for schedule_invs
+   will be simpler with it *)
```

Comment:
hang on, strengthening a precondition makes the proof simpler?!

Reply:
Good point, no idea what the intention was behind this comment originally but I was able to remove that precondition and the proofs still went through.